docs/question/server.md

```diff
@@ -1,89 +1,312 @@
 
-<!-- y is defined by data["correct_answers"]["y"] in server.py's `generate()`. -->
+You can set `data["format_errors"][NAME]` to mark the submission as invalid. This will cause the question to not use up one of the student's attempts on assessments. However, you are encouraged to do as many checks for invalid data as possible in `parse` instead of `grade`; the `parse` function is called when the student hits "Save only", in manually graded questions, and in assessments without real-time grading.
+
+```python title="server.py"
+import math
+import prairielearn as pl
+
+def grade(data):
+    # Give half points for incorrect answers larger than "x", only if not already correct.
+    # Use math.isclose to avoid possible floating point errors.
+    y_is_correct = math.isclose(data["partial_scores"]["y"]["score"], 1.0)
+    if not y_is_correct and data["submitted_answers"]["y"] > data["params"]["x"]:
+        data["partial_scores"]["y"]["score"] = 0.5
+        pl.set_weighted_score_data(data)
+        data["feedback"]["y"] = "Your value for $y$ is larger than $x$, but incorrect."
+```
+
+### Providing feedback
+
+To set custom feedback, the grading function should set the corresponding entry in the `data["feedback"]` dictionary. These feedback entries are passed in when rendering the `question.html`, which can be accessed by using the mustache prefix `{{feedback.}}`. See the [above example](#complete-example) or [this demo question](https://github.com/PrairieLearn/PrairieLearn/tree/master/exampleCourse/questions/demo/custom/gradeFunction) for examples of this.
+
+Some elements provide feedback through `data["partial_scores"][NAME]["feedback"]`, which you can override in the grading function. This field is a string, except for the `<pl-drawing>` and `<pl-checkbox>` elements. The feedback provided here will be attached to the student's answer for that element, which can make it easier for students to interpret the feedback they received in longer questions. Since not elements provide feedback this way, and use different data structures to represent the feedback they give, we recommend using `data["feedback"][NAME]` to provide question-specific feedback instead.
+
+!!! tip "Answer-specific feedback"
+
+    If you want to provide feedback on a specific answer that doesn't use/support `data["partial_scores"][NAME]["feedback"]`,
+    you could render the feedback directly after the element in `question.html`:
+
+    <!-- prettier-ignore -->
+    ```html title="question.html"
+    <pl-string-input ...></pl-string-input>
+    {{#feedback.value1}}<div>Feedback: {{.}}</div>{{/feedback.value1}}
+    ```
+
+Overall question feedback in `data["feedback"][NAME]` needs be rendered explicity in the `question.html` template using `{{feedback.NAME}}`. Feedback given in `data["partial_scores"][NAME]["feedback"]` will be rendered automatically by the elements, assuming the element is visible.
```

Comment:
```suggestion
Overall question feedback in `data["feedback"][NAME]` needs be rendered explicity in the `question.html` template using `{{feedback.NAME}}`. Feedback given in `data["partial_scores"][NAME]["feedback"]`, if supported by the element, will be rendered automatically by the elements, assuming the element is visible.
```

Reply:
Is this redundant with the paragraph above? "Some elements provide feedback through..." and "Since not elements provide feedback this way,"

Reply:
It's not redundant, it's consistent. As written it implies the feedback is the norm, not a limited feature.

Reply:
Makes sense, I had intended for a different interpretation but I didn't consider that one. You change was applied in [70ac407](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/70ac4076901e51062960fa1123b5d56f95a12a4b).